lvmd/command/lvm.go

```diff
@@ -1,125 +1,87 @@
 package command
 
 import (
-	"bytes"
 	"context"
 	"errors"
 	"fmt"
-	"os"
-	"os/exec"
 	"path"
-
-	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
 const (
 	nsenter = "/usr/bin/nsenter"
 	lvm     = "/sbin/lvm"
 )
 
-var Containerized bool = false
-
 // ErrNotFound is returned when a VG or LV is not found.
 var ErrNotFound = errors.New("not found")
 
-// wrapExecCommand calls cmd with args but wrapped to run
-// on the host
-func wrapExecCommand(cmd string, args ...string) *exec.Cmd {
-	if Containerized {
-		args = append([]string{"-m", "-u", "-i", "-n", "-p", "-t", "1", cmd}, args...)
-		cmd = nsenter
-	}
-	c := exec.Command(cmd, args...)
-	return c
-}
-
-// callLVM calls lvm sub-commands.
-// cmd is a name of sub-command.
-func callLVM(ctx context.Context, cmd string, args ...string) error {
-	_, err := callLVMWithStdout(ctx, cmd, args...)
-	return err
-}
-
-// callLVMWithStdout calls lvm sub-commands and returns stdout.
-// cmd is a name of sub-command.
-func callLVMWithStdout(ctx context.Context, cmd string, args ...string) ([]byte, error) {
-	var stdout bytes.Buffer
-	args = append([]string{cmd}, args...)
-
-	c := wrapExecCommand(lvm, args...)
-	c.Env = os.Environ()
-	c.Env = append(c.Env, "LC_ALL=C")
-	c.Stdout = &stdout
-	c.Stderr = os.Stderr
-
-	log.FromContext(ctx).Info("invoking LVM command", "args", args)
-	err := c.Run()
-	return stdout.Bytes(), err
-}
-
-// LVInfo is a map of lv attributes to values.
-type LVInfo map[string]string
-
 // VolumeGroup represents a volume group of linux lvm.
+// The state should be considered immutable and will not automatically update.
+// The Update method should be called to refresh the state in case it is known that the state may have changed.
 type VolumeGroup struct {
 	state vg
-	lvs   []lv
+	// reportLvs use with getLVMState, which populates vg and lv at the same time.
+	// should not be used otherwise as fields are fetched dynamically.
+	reportLvs map[string]lv
+}
+
+// getLVMState returns the current state of lvm lvs for the given volume group.
+// If lvname is empty, all lvs are returned. Otherwise, only the lv with the given name is returned or an error if not found.
+func getLVs(ctx context.Context, vg *VolumeGroup, lvname string) (map[string]lv, error) {
+	// use fast path if we have the lvs already through the report
+	if len(vg.reportLvs) > 0 {
+		if lvname != "" {
+			if lvFromMap, ok := vg.reportLvs[lvname]; ok {
+				return map[string]lv{lvname: lvFromMap}, nil
+			}
+			return nil, ErrNotFound
+		}
+		return vg.reportLvs, nil
+	}
+
+	// by default, fetch all lvs for the vg
+	name := vg.state.name
+	// if lvname is set, only fetch that lv in the vg
+	if lvname != "" {
+		name += "/" + lvname
+	}
+
+	return getLVReport(ctx, name)
```

Comment:
I also thought about this but ultimately decided on this approach.

This is because the caching is currently done only if the lvm state report is used (which contains both lv and vg). This is because otherwise we would have to blindly update the LVs and VGs again with Update() whenever the cache becomes inconsistent. (e.g. one lv update would have to invalidate the report, or we would have to write a smarter logic that only invalidates that lv and then fetches it again the next time it is used).

After verifying all usages of lvm.go functions in lv/vgservice, I concluded we never make use of them, and while this is in theory a good addition, in practice the cache becomes irrelevant.

If you find a call case where we do not use lvmstate and fetch lvs without name multiple times, please let me know!